### PR TITLE
pass args to scripts

### PR DESCRIPTION
**Pull request by Yakiyo**
Pass any additional arguments passed after script name in command invocation to the script.

Resolves https://github.com/pnpm/pn/issues/48

### CODE REVIEW

**Comment by Yakiyo** [comment]
Any status on this?

File: src/main.rs [hunk 1]
```diff
@@ -73,12 +73,21 @@ fn run() -> Result<(), MainError> {
         eprintln!("> {command}\n");
         run_script(name, command, cwd)
     };
+    let command_string = |name: &str, args: Vec<String>| {
+        // not sure creating vec with cap makes much diff here
```
Comment (zkochan) [inline src/main.rs:77]: This comment should be removed

Comment (Yakiyo) [reply src/main.rs]: done

File: src/main.rs [hunk 2]
```diff
@@ -73,12 +73,20 @@ fn run() -> Result<(), MainError> {
         eprintln!("> {command}\n");
         run_script(name, command, cwd)
     };
+    let command_string = |name: &str, args: Vec<String>| {
+        let mut commands = Vec::with_capacity(args.len() + 1);
+        commands.push(name.to_string());
+        let mut args = args;
+        commands.append(&mut args);
+        commands.join(" ")
+    };
```
Comment (KSXGitHub) [inline src/main.rs:82]: Is it really necessary to allocate a whole `Vec` just to join a string?

```suggestion
    let command_string = |name: &str, args: &[String]| {
        let mut string = name.to_string();
        for arg in args {
            string += " ";
            string += arg;
        }
        string
    };
```

File: src/main.rs [hunk 3]
```diff
@@ -100,7 +108,8 @@ fn run() -> Result<(), MainError> {
                     return pass_to_pnpm(&args); // args already contain name, no need to prepend
                 }
                 if let Some(command) = manifest.scripts.get(name) {
-                    return print_and_run_script(&manifest, name, command, &cwd);
+                    let commands = command_string(command, args[1..].to_vec());
```
Comment (KSXGitHub) [inline src/main.rs:111]: Why `.to_vec()`?

**Review by KSXGitHub — CHANGES_REQUESTED** [review]

File: src/main.rs [hunk 4]
```diff
@@ -73,12 +73,20 @@ fn run() -> Result<(), MainError> {
...
+        commands.join(" ")
+    };
     match cli.command {
         cli::Command::Run(args) => {
             let (cwd, manifest) = cwd_and_manifest()?;
             if let Some(name) = args.script {
                 if let Some(command) = manifest.scripts.get(&name) {
-                    print_and_run_script(&manifest, &name, command, &cwd)
+                    let commands = command_string(command, args.args);
+                    print_and_run_script(&manifest, &name, &commands, &cwd)
```
Comment (KSXGitHub) [inline src/main.rs:89]: Why plural?

```suggestion
                    let command = command_string(command, args.args);
                    print_and_run_script(&manifest, &name, &command, &cwd)
```

You don't need to be concerned about name collision, Rust will shadow the name.

File: src/main.rs [hunk 5]
```diff
@@ -100,7 +108,8 @@ fn run() -> Result<(), MainError> {
                     return pass_to_pnpm(&args); // args already contain name, no need to prepend
                 }
                 if let Some(command) = manifest.scripts.get(name) {
-                    return print_and_run_script(&manifest, name, command, &cwd);
+                    let commands = command_string(command, args[1..].to_vec());
+                    return print_and_run_script(&manifest, name, &commands, &cwd);
```
Comment (KSXGitHub) [inline src/main.rs:112]: Please remove the plural here also.

Comment (KSXGitHub) [reply src/main.rs:112]:
```suggestion
                    let command = command_string(command, args[1..].to_vec());
                    return print_and_run_script(&manifest, name, &command, &cwd);
```

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
What about the arguments that contain whitespaces? Don't we need to escape them before passing the command to `sh -c`?

Related crates: https://docs.rs/shell-escape/0.1.5/shell_escape/

**Comment by Yakiyo** [comment]
> What about the arguments that contain whitespaces? Don't we need to escape them before passing the command to `sh -c`?
> 
> Related crates: https://docs.rs/shell-escape/0.1.5/shell_escape/

Not sure what that would look like. care to give an example? How does pnpm do it?

**Comment by KSXGitHub** [comment]
> Not sure what that would look like. care to give an example? How does pnpm do it?

pnpm for sure handles whitespace.

proof

But for pnpm did it, you must ask @zkochan.

If you ask me, I have already answered: Just use the crate I suggested, I don't really care how exactly pnpm did it.

**Comment by Yakiyo** [comment]
> pnpm for sure handles whitespace.
> 

i tried it out a few seconds ago but clap handles it. 
passing `pn run test "Hello World" "! !"` takes it as two args instead of 4.

**Comment by Yakiyo** [comment]
okay i understand the issue now
image

**Comment by KSXGitHub** [comment]
If you used my [list-args.js](https://user-images.githubusercontent.com/11488886/258154159-fbe54b8e-2de3-4527-8932-7e3e22e244cf.png), the issue would have been clearer.

Also, please use the `shell-quote` crate as suggested [here](https://github.com/pnpm/pn/pull/49#issuecomment-1664169793).

**Comment by Yakiyo** [comment]
@KSXGitHub not sure if i did anything wrong but it doesnt seem to work

image

and based on the ci, shell quote is failing to build on windows.

**Comment by Yakiyo** [comment]
`env::args_os` itself is stripping away the quotes
image

**Comment by KSXGitHub** [comment]
> shell quote is failing to build on windows.

This was an oversight on my part. I was not aware that shell-quote is not cross-platform. Let's try [os_display](https://docs.rs/os_display/0.1.3/os_display/).

Also, changing the type of `RunArgs::args` to `Vec<OsString>` seems to lead to verbose code, let's reverse that.

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
See https://github.com/Yakiyo/pn/pull/1

**Comment by Yakiyo** [comment]
> See [Yakiyo#1](https://github.com/Yakiyo/pn/pull/1)

merged

**Comment by KSXGitHub** [comment]
> Windows CI fails

CLRF strikes again.

File: tests/test_main.rs [hunk 6]
```diff
@@ -46,6 +46,42 @@ fn run_script() {
...
+        .arg("run")
+        .arg("list-args")
+        .arg("foo")
+        .arg("bar")
+        .arg("hello world")
+        .arg("! !@")
+        .arg("abc def ghi")
+        .assert()
+        .success()
+        .stdout(include_str!("fixtures/append-script-args/stdout.txt"));
```
Comment (KSXGitHub) [inline tests/test_main.rs:84]:
```suggestion
        .stdout(include_str!("fixtures/append-script-args/stdout.txt").replace("\r\n", "\n"));
```

**Review by KSXGitHub — APPROVED** [review]

**Review by KSXGitHub — CHANGES_REQUESTED** [review]
Can you merge this? https://github.com/Yakiyo/pn/pull/2

**Comment by Yakiyo** [comment]
> Can you merge this? [Yakiyo#2](https://github.com/Yakiyo/pn/pull/2)

done

**Review by zkochan — APPROVED** [review]

**Pull request merged**